core/src/com/fourinachamber/fortyfive/map/shop/ShopCardsHandler.kt

```diff
@@ -0,0 +1,169 @@
+        labels.add(label)
+    }
+
+    private fun getCardToAddWithChances(rnd: Random): Int {
+        val maxWeight = chances.map { (_, b) -> b }.sum()
+        val value = (0.0F..maxWeight).random(rnd)
+        var curSum = 0.0
+        for (e in chances) {
+            curSum += e.value
+            if (curSum > value) {
+                return _allCards.indexOf(_allCards.find { it.name == e.key })
+            }
+        }
+        return chances.size - 1
+    }
+
+    fun calculateChances(
+        type: String,
+        shopFile: OnjObject,
+        person: OnjObject
+    ) { //TODO biome, when they are added
+        val allTypes = shopFile.get<OnjArray>("types").value.map { it as OnjObject }
+        val curTypeChances = if (type !in allTypes.map { it.get<String>("name") }) {
+            allTypes.first { it.get<String>("name") == person.get<String>("defaultShopParameter") }
+                .get<OnjArray>("cardChanges").value.map { it as OnjObject }
+        } else {
+            allTypes.first { it.get<String>("name") == type }.get<OnjArray>("cardChanges").value.map { it as OnjObject }
+        }
+        _allCards.forEach { chances[it.name] = 0F }
+        curTypeChances.forEach {
+            applyChancesEffect(
+                it.get<OnjNamedObject>("select"),
+                it.get<OnjNamedObject>("effect")
+            )
+        }
+    }
+
+    private fun applyChancesEffect(selector: OnjNamedObject, effect: OnjNamedObject) {
+        val cardsToChange: List<String> = _allCards.filter {
+            (if (selector.name == "ByName") it.name == selector.get<String>("name") else it.tags.contains(
+                selector.get<String>("name")
+            ))
+        }.map { it.name }
+        when (effect.name) {
+            "Blacklist" -> {
+                cardsToChange.forEach { chances.remove(it) }
+            }
+
+            "ProbabilityAddition" -> {
+                chances.map { (a, _) -> a }.filter { it in cardsToChange }.forEach {
+                    chances[it] = chances[it]!! + effect.get<Double>("weight").toFloat()
+                }
+            }
+
+            "PriceMultiplier" -> {
+                _allCards.filter { it.name in cardsToChange }.forEach {
+                    it.price = (it.price * effect.get<Double>("price")).toInt()
+                }
+            }
+        }
+    }
+
+    fun buyCard(
+        cardImg: CustomImageActor,
+    ) {
+        val i = cardWidgets.indexOf(cardImg)
+        if (i !in boughtIndices) boughtIndices.add(i)
+        val card = cards[i]
+        SaveState.playerMoney -= card.price
+        updateCards()
+        FortyFiveLogger.debug(logTag, "Bought ${card.name} for a price of ${card.price}")
+        SaveState.buyCard(card.name)
+    }
+
+    private fun updateCards() {
+        for (i in cardWidgets.indices) {
+            if (i in boughtIndices) {
+                labels[i].setText("bought")
+                cardWidgets[i].styleManager?.enterActorState("unbuyable")
```

Comment:
i think it would be convinient to have extension functions on StyledActor for things like entering, leaving or checking states